text/dependencies/rfcs/0000-dependency-management-phase-one.md

```diff
@@ -0,0 +1,440 @@
+      discovered at the `output` path. There may be multiple entries per
+      version, corresponding to different target OS variants of a
+      dependency. The metadata must adhere to the following schema in order
+      to work as expected with a generic tool for adding metaata to the
+      `buildpack.toml`.
+      <details>
+      <summary>Schema</summary>
+
+      ```
+      {
+        "name": <dependency name>,
+        "version": <version>,
+        "sha256": <SHA256>,
+        "uri": <URI of dependency>,
+        "stacks": [{"id": <compatible stack ID> }],
+        "source": <source URI>,
+        "source_sha256": <SHA256 of source dependency>,
+        "deprecation_date": <optional field, deprecation date>,
+        "cpe": <CPE>,
+        "purl": <package URL>,
+        "licenses": [ <list of licenses, by ID> ],
+        "target": <generic name of the variant, to be used during compilation>
+      }
+
+      ```
+      </details>
+#### Transfer/Combine Version Retrieval and Metadata Generation
+
+The new  version retrieval code will merge together the job of discovering
+versions and getting metadata for each version. It will pick up all newer
+versions than what the `buildpack.toml` contains, and outputs a JSON file with
+metadata entries for each dependency version discovered. This code should be
+ported from the dep-server repository to the buildpack under a directory named
+`dependency/retrieval`, taking in the path to the `buildpack.toml` file and the
+path to the output file. It should be well-documented and useable on a local
+environment. The location must be standardized for use in automation.
+
+Buildpack authors should feel free to reuse (or rewrite) the code from the
+dep-server for these parts as they see fit. The language that it is written in is
+up to maintainers; however, if written in Go it will be easier to leverage
+future abstraction libraries that may arise.
+
+Version discovery code comes from
+[dep-server/pkg/dependency](https://github.com/paketo-buildpacks/dep-server/tree/main/pkg/dependency).
+Each dependency has a slightly different way new versions are made
+discoverable. The code will use the `buildpack.toml` dependency constraint
+metadata and versions in the `buildpack.toml` already to determine new
+versions.
+
+Metadata generation code also comes from
+[dep-server/pkg/dependency](https://github.com/paketo-buildpacks/dep-server/tree/main/pkg/dependency).
+Code to generate/gather all of the metadata for a dependency should be moved and combined
+into the retrieval code.  The code should
+do essentially the same thing that the existent code does, and support the same
+metadata fields.
+The supported metadata fields are: `source URI`, `source SHA256`, `version`,
+`URI`, `SHA256`, `ReleaseDate`, `DeprecationDate`, `stacks`, `purl`, `licenses`
+and `CPE`.
+
+#### Separating OS Variants and Supporting Multiple Stacks
+An additional metaadata field named `target` will be added, which will be the
```

Comment:
A couple of concerns about this:

1. At the CNB level, there has been talk for a while about removing the concept of a "stack". See [RFC 96](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md). There are a number of references to stack in this proposal, in particular this section. I think we should try to get ahead of that change so we're not coupling this process to a term that's being phased out. I think we should try to use [this terminology instead](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md#base-image-metadata).

2. Somewhat related, but I think we need to include support for multiple architectures in this proposal. ARM64 support is hopefully not that far out and dependency support for multiple architectures is a requirement for that. I'm not sure if `target` was intended to cover architecture as well. I could see this being `target` being a format like `ARCH-VENDOR-SYS-ABI` ex: `aarch64-apple-darwin` or `x86_64-unknown-linux-gnu` or `x86_64-unknown-linux-musl`. I could also see having individual properties for each of those value, like `arch=x86_64`, `vendor=ubuntu-bionic`, `sys=linux`, `abi=musl` (not sure we need all of these fields, names could change as well, just an example).

Reply:
@dmikusa-pivotal I have been aware that stacks will go away in the future- beyond the specific language that calls out "stacks" in the RFC, I actually think that this proposal should still work more or less for the new world with the new fields, without any major architectural changes. I think it should be fine to move forward with the proposal as is and then when we actually get to a place in the project where we remove stacks - we can make amendments as needed.

Reply:
Discussed in WG meeting.

There's no reason to believe that deferring this work and removing stack terminology to a later date will be an overly difficult task (we can add/deprecate fields in metadata and adjust implementations). There is also concern that trying to do it now might create challenges as the CNB project hasn't implemented this yet, so there's still a chance for things to change/deviate from the RFC. Thus the consensus on the call was to proceed now with the stack terminology.